state-sync/diem-data-client/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+    #[error("The requested data is too large: {0}")]
+    DataIsTooLarge(String),
+    #[error("Timed out waiting for a response: {0}")]
+    TimeoutWaitingForResponse(String),
+    #[error("Unexpected error encountered: {0}")]
+    UnexpectedErrorEncountered(String),
+}
+
+/// A response error that users of the Diem Data Client can use to notify
+/// the Data Client about invalid or malformed responses.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum ResponseError {
+    InvalidData,
+    MissingData,
+    ProofVerificationError,
+}
+
+/// The API offered by the Diem Data Client.
+#[async_trait]
+pub trait DiemDataClient {
+    /// Returns a single account states chunk with proof, containing the accounts
+    /// from start to end index (inclusive) at the specified version. The proof
+    /// version is the same as the specified version.
+    async fn get_account_states_with_proof(
+        &self,
+        version: u64,
+        start_index: HashValue,
+        end_index: HashValue,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Returns all epoch ending ledger infos between start and end (inclusive).
+    /// If the data cannot be fetched (e.g., the number of epochs is too large),
+    /// an error is returned.
+    async fn get_epoch_ending_ledger_infos(
+        &self,
+        start_epoch: u64,
+        end_epoch: u64,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Returns a global summary of the data currently available in the network.
+    fn get_global_data_summary(&self) -> Result<DataClientResponse, Error>;
+
+    /// Returns the number of account states at the specified version.
+    async fn get_number_of_account_states(&self, version: u64)
+        -> Result<DataClientResponse, Error>;
+
+    /// Returns a transaction output list with proof object, with transaction
+    /// outputs from start to end versions (inclusive). The proof is relative to
+    /// the specified `proof_version`. If the data cannot be fetched (e.g., the
+    /// number of transaction outputs is too large), an error is returned.
+    async fn get_transaction_outputs_with_proof(
+        &self,
+        proof_version: u64,
+        start_version: u64,
+        end_version: u64,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Returns a transaction list with proof object, with transactions from
+    /// start to end versions (inclusive). The proof is relative to the specified
+    /// `proof_version`. If the data cannot be fetched (e.g., the number of
+    /// transactions is too large), an error is returned.
+    async fn get_transactions_with_proof(
+        &self,
+        proof_version: u64,
+        start_version: u64,
+        end_version: u64,
+    ) -> Result<DataClientResponse, Error>;
+
+    /// Notifies the Diem Data Client about a previously received response that
+    /// was bad (e.g., invalid or malformed).
+    async fn notify_bad_response(
+        &self,
+        response_id: u64,
```

Comment:
might be missing something but what value is a response id if we don't know the request? should this be request id? or both?

Reply:
The `DiemDataClient` will return a unique response id for each API request, e.g., `get_transactions_with_proof()`. Internally, the `DiemDataClient` will maintain a (limited) history regarding which peer provided the response that maps to the corresponding response id. So, using this call, the `DiemDataClient` will penalize the peer who gave the bad response. Not sure if this answers your question?